### DISCUSSION TITLE
Migrating a library to SQLAlchemy 2.0, however one sample test failing due to AttributeError: no attribute _active_history

### DISCUSSION THREAD

**Discussion by kiraksi** [sqlalchemy/sqlalchemy#10743] [Usage Questions] [answered]
Hi all, I am trying to make a sqlalchemy connector library compatible with sqlalchemy 2.0, however one of my sample tests fail with an AttributeError:

```def
from . import STRUCT

sixeses = STRUCT.example(engine)

STRUCT\_test.py:24:

---

STRUCT.py:50: in example
sebring = Car(model="Sebring",
\<string>:4: in **init**
???
.nox/py-3-11/lib/python3.11/site-packages/sqlalchemy/orm/state.py:566: in \_initialize\_instance
with util.safe\_reraise():
.nox/py-3-11/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py:146: in **exit**
raise exc\_value.with\_traceback(exc\_tb)
.nox/py-3-11/lib/python3.11/site-packages/sqlalchemy/orm/state.py:564: in \_initialize\_instance
manager.original\_init(\*mixed\[1:\], \*\*kwargs)
.nox/py-3-11/lib/python3.11/site-packages/sqlalchemy/orm/decl\_base.py:2137: in \_declarative\_constructor
setattr(self, k, kwargs\[k\])
.nox/py-3-11/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py:537: in **set**
self.impl.set(

---

self = \<sqlalchemy.orm.attributes.ScalarAttributeImpl object at 0x10a0f3880>
state = \<sqlalchemy.orm.state.InstanceState object at 0x10a0f49b0>
dict\_ = {'\_sa\_instance\_state': \<sqlalchemy.orm.state.InstanceState object at 0x10a0f49b0>, 'model': 'Sebring'}
value = {'cylinder': {'compression': 18.0, 'count': 6}, 'horsepower': 235}, initiator = None, passive = symbol('PASSIVE\_OFF')
check\_old = None, pop = False

def set(
self,
state: InstanceState\[Any\],
dict\_: Dict\[str, Any\],
value: Any,
initiator: Optional\[AttributeEventToken\] = None,
passive: PassiveFlag = PASSIVE\_OFF,
check\_old: Optional\[object\] = None,
pop: bool = False,
) -> None:

if self.dispatch.\_active\_history:

E       AttributeError: 'NoneType' object has no attribute '\_active\_history
```

In order to create our Struct datatype we use a kind of hacky way to access metadata that worked in 1.4 and earlier version, but only breaks for a specific test in 2.0 with this AttributeError. Im not sure if its a sqlalchemy issue or an issue with our implementation and would love some guidance

Struct Implementation:

```py
import packaging.version
import sqlalchemy.sql.default_comparator
import sqlalchemy.sql.sqltypes
import sqlalchemy.types

from . import base

sqlalchemy_1_4_or_more = packaging.version.parse(
    sqlalchemy.__version__
) >= packaging.version.parse("1.4")

if sqlalchemy_1_4_or_more:
    import sqlalchemy.sql.coercions
    import sqlalchemy.sql.roles


def _get_subtype_col_spec(type_):
    global _get_subtype_col_spec

    type_compiler = base.dialect.type_compiler(base.dialect())
    _get_subtype_col_spec = type_compiler.process
    return _get_subtype_col_spec(type_)


class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):


    def __init__(
        self,
        *fields,
        **kwfields,
    ):
        # Note that because:
        # https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order
        # We know that `kwfields` preserves order.
        self._STRUCT_fields = tuple(
            (
                name,
                type_ if isinstance(type_, sqlalchemy.types.TypeEngine) else type_(),
            )
            for (name, type_) in (fields + tuple(kwfields.items()))
        )

        self._STRUCT_byname = {
            name.lower(): type_ for (name, type_) in self._STRUCT_fields
        }

    def __repr__(self):
        fields = ", ".join(
            f"{name}={repr(type_)}" for name, type_ in self._STRUCT_fields
        )
        return f"STRUCT({fields})"

    def get_col_spec(self, **kw):
        fields = ", ".join(
            f"{name} {_get_subtype_col_spec(type_)}"
            for name, type_ in self._STRUCT_fields
        )
        return f"STRUCT<{fields}>"

    def bind_processor(self, dialect):
        return dict

    class Comparator(sqlalchemy.sql.sqltypes.Indexable.Comparator):
        def _setup_getitem(self, name):
            if not isinstance(name, str):
                raise TypeError(
                    f"STRUCT fields can only be accessed with strings field names,"
                    f" not {repr(name)}."
                )
            subtype = self.expr.type._STRUCT_byname.get(name.lower())
            if subtype is None:
                raise KeyError(name)
            operator = struct_getitem_op
            index = _field_index(self, name, operator)
            return operator, index, subtype

        def __getattr__(self, name):
            if name.lower() in self.expr.type._STRUCT_byname:
                return self[name]

    comparator_factory = Comparator


# In the implementations of _field_index below, we're stealing from
# the JSON type implementation, but the code to steal changed in
# 1.4. :/

if sqlalchemy_1_4_or_more:

    def _field_index(self, name, operator):
        return sqlalchemy.sql.coercions.expect(
            sqlalchemy.sql.roles.BinaryElementRole,
            name,
            expr=self.expr,
            operator=operator,
            bindparam_type=sqlalchemy.types.String(),
        )

else:

    def _field_index(self, name, operator):
        return sqlalchemy.sql.default_comparator._check_literal(
            self.expr,
            operator,
            name,
            bindparam_type=sqlalchemy.types.String(),
        )


def struct_getitem_op(a, b):
    raise NotImplementedError()


sqlalchemy.sql.default_comparator.operator_lookup[
    struct_getitem_op.__name__
] = sqlalchemy.sql.default_comparator.operator_lookup["json_getitem_op"]


class SQLCompiler:
    def visit_struct_getitem_op_binary(self, binary, operator_, **kw):
        left = self.process(binary.left, **kw)
        return f"{left}.{binary.right.value}"


```

**Comment by CaselIT** [comment]
Hi,

The `class STRUCT` does not seem connected with the error.
Can you share the code that raises the error and the definition of `Car`?

**Reply by kiraksi** [reply]
Yes, here is how struct is utilized:
```py
def example(engine):
    # fmt: off
    # [START bigquery_sqlalchemy_create_table_with_struct]
    from sqlalchemy.ext.declarative import declarative_base
    from sqlalchemy import Column, String, Integer, Float
    from sqlalchemy_bigquery import STRUCT

    Base = declarative_base()

    class Car(Base):
        __tablename__ = "Cars"

        model = Column(String, primary_key=True)
        engine = Column(
            STRUCT(
                cylinder=STRUCT(("count", Integer),
                                ("compression", Float)),
                horsepower=Integer)
            )

    # [END bigquery_sqlalchemy_create_table_with_struct]
    Car.__table__.create(engine)

    # [START bigquery_sqlalchemy_insert_struct]
    from sqlalchemy.orm import sessionmaker

    Session = sessionmaker(bind=engine)
    session = Session()

    sebring = Car(model="Sebring",
                  engine=dict(
                      cylinder=dict(
                          count=6,
                          compression=18.0),
                      horsepower=235))
    townc = Car(model="Town and Counttry",
                engine=dict(
                    cylinder=dict(
                        count=6,
                        compression=16.0),
                    horsepower=251))
    xj8 = Car(model="XJ8",
              engine=dict(
                  cylinder=dict(
                      count=8,
                      compression=10.75),
                  horsepower=575))

    session.add_all((sebring, townc, xj8))
    session.commit()

    # [END bigquery_sqlalchemy_insert_struct]

    # [START bigquery_sqlalchemy_query_struct]
    sixes = session.query(Car).filter(Car.engine.cylinder.count == 6)
    # [END bigquery_sqlalchemy_query_struct]
    sixes1 = list(sixes)

    # [START bigquery_sqlalchemy_query_STRUCT]
    sixes = session.query(Car).filter(Car.engine.CYLINDER.COUNT == 6)
    # [END bigquery_sqlalchemy_query_STRUCT]
    sixes2 = list(sixes)

    # [START bigquery_sqlalchemy_query_getitem]
    sixes = session.query(Car).filter(Car.engine["cylinder"]["count"] == 6)
    # [END bigquery_sqlalchemy_query_getitem]
    # fmt: on
    sixes3 = list(sixes)

    return sixes1, sixes2, sixes3

def test_struct(engine):
    from . import STRUCT

    sixeses = STRUCT.example(engine)

    for sixes in sixeses:
        assert sorted(car.model for car in sixes) == ["Sebring", "Town and Counttry"]
```

**Reply by CaselIT** [reply]
Can you try calling https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.configure_mappers after creating the car table? (ie just after `Car.__table__.create(engine)`

Other than that I don't many other suggestions at the moment. @zzzeek does anything seems to jump out?

**Reply by kiraksi** [reply]
I tried calling it however, I still get the same attribute error. I was looking at the difference between active_history in 1.4 vs 2.0, and it seems like a major change is that now its enabled by default? Could that have something to do with it?

**Reply by CaselIT** [reply]
Thanks for trying. It's strange since `dispatch` should never be None, so an error there is very strange.

Just to try excluding simple stuff first, could you try uninstalling and re-installing sqlalchemy just to ensure that no changes were done to it??

**Reply by kiraksi** [reply]
Yes I tried uninstalling and reinstalling sqlalchemy, I also tried installing different versions of the 2.0 release up till 2.0.22. And yes I dont believe dispatch is even included in our code, which confuses me as well since there can be no instance where its changed, which led me to thinking that maybe this is a bug in the sqlalchemy library? I don't really know where else the error can come from then, but thats my simple understanding.

**Comment by zzzeek** [comment]
the issue has to do with the ORM object attribute not fully setting itself up for some reason.    If the only odd thing is a custom type with a custom comparator, probably something about the comparator class is causing some kind of problem, but strange that it would come out in that way.

since you have code for your custom type above can you organize it into a full example with an ORM mapping that reproduces the issue?   If you can make an all-in-one-paste reproducer script I can find the problem quickly.

**Reply by kiraksi** [reply]
It's a bit complicated to make this a produceable script since it involves code from multiple files, but this is actually the python-bigquery-sqlalchemy repository I am working on. Specifically PR https://github.com/googleapis/python-bigquery-sqlalchemy/pull/920. If you go to that branch and pull the current changes, you can cd into samples/snippets and run pytest on the struct_test.py file to reproduce this issue, will that work for you? I just pushed my current changes. If not I can try to make a very long script but this is probably the easiest way to reproduce the issue I am having. Thank you!

**Reply by zzzeek** [reply]
OK I guess the ORM test is the one you showed above [here](https://github.com/sqlalchemy/sqlalchemy/discussions/10743#discussioncomment-7793944)

**Comment by zzzeek** [comment]
OK you, and we, are super lucky here, because the code above adapted to SQLite and taking out the create tables part, since this is just the ORM setup before anything happens, reproduces the problem, which is quite unusual to hit it so soon.   Here's a script that when run on 2.0 will show this problem:


```py
import sqlalchemy.sql.default_comparator
import sqlalchemy.sql.sqltypes
import sqlalchemy.types

import sqlalchemy.sql.coercions
import sqlalchemy.sql.roles

from sqlalchemy.dialects.sqlite import base


def _get_subtype_col_spec(type_):
    global _get_subtype_col_spec

    type_compiler = base.SQLiteDialect.type_compiler_cls(base.dialect())
    _get_subtype_col_spec = type_compiler.process
    return _get_subtype_col_spec(type_)


class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):


    def __init__(
        self,
        *fields,
        **kwfields,
    ):
        # Note that because:
        # https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order
        # We know that `kwfields` preserves order.
        self._STRUCT_fields = tuple(
            (
                name,
                type_ if isinstance(type_, sqlalchemy.types.TypeEngine) else type_(),
            )
            for (name, type_) in (fields + tuple(kwfields.items()))
        )

        self._STRUCT_byname = {
            name.lower(): type_ for (name, type_) in self._STRUCT_fields
        }

    def __repr__(self):
        fields = ", ".join(
            f"{name}={repr(type_)}" for name, type_ in self._STRUCT_fields
        )
        return f"STRUCT({fields})"

    def get_col_spec(self, **kw):
        fields = ", ".join(
            f"{name} {_get_subtype_col_spec(type_)}"
            for name, type_ in self._STRUCT_fields
        )
        return f"STRUCT<{fields}>"

    def bind_processor(self, dialect):
        return dict

    class Comparator(sqlalchemy.sql.sqltypes.Indexable.Comparator):
        def _setup_getitem(self, name):
            if not isinstance(name, str):
                raise TypeError(
                    f"STRUCT fields can only be accessed with strings field names,"
                    f" not {repr(name)}."
                )
            subtype = self.expr.type._STRUCT_byname.get(name.lower())
            if subtype is None:
                raise KeyError(name)
            operator = struct_getitem_op
            index = _field_index(self, name, operator)
            return operator, index, subtype

        def __getattr__(self, name):
            if name.lower() in self.expr.type._STRUCT_byname:
                return self[name]

    comparator_factory = Comparator


# In the implementations of _field_index below, we're stealing from
# the JSON type implementation, but the code to steal changed in
# 1.4. :/


def _field_index(self, name, operator):
    return sqlalchemy.sql.coercions.expect(
        sqlalchemy.sql.roles.BinaryElementRole,
        name,
        expr=self.expr,
        operator=operator,
        bindparam_type=sqlalchemy.types.String(),
    )



def struct_getitem_op(a, b):
    raise NotImplementedError()


sqlalchemy.sql.default_comparator.operator_lookup[
    struct_getitem_op.__name__
] = sqlalchemy.sql.default_comparator.operator_lookup["json_getitem_op"]


class SQLCompiler:
    def visit_struct_getitem_op_binary(self, binary, operator_, **kw):
        left = self.process(binary.left, **kw)
        return f"{left}.{binary.right.value}"




from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, String, Integer, Float, create_engine

Base = declarative_base()
engine = create_engine("sqlite://")

class Car(Base):
    __tablename__ = "Cars"

    model = Column(String, primary_key=True)
    engine = Column(
        STRUCT(
            cylinder=STRUCT(("count", Integer),
                            ("compression", Float)),
            horsepower=Integer)
        )


# [START bigquery_sqlalchemy_insert_struct]
from sqlalchemy.orm import sessionmaker

Session = sessionmaker(bind=engine)
session = Session()

sebring = Car(model="Sebring",
                engine=dict(
                    cylinder=dict(
                        count=6,
                        compression=18.0),
                    horsepower=235))
townc = Car(model="Town and Counttry",
            engine=dict(
                cylinder=dict(
                    count=6,
                    compression=16.0),
                horsepower=251))
xj8 = Car(model="XJ8",
            engine=dict(
                cylinder=dict(
                    count=8,
                    compression=10.75),
                horsepower=575))

session.add_all((sebring, townc, xj8))
session.commit()

# [END bigquery_sqlalchemy_insert_struct]

# [START bigquery_sqlalchemy_query_struct]
sixes = session.query(Car).filter(Car.engine.cylinder.count == 6)
# [END bigquery_sqlalchemy_query_struct]
sixes1 = list(sixes)

# [START bigquery_sqlalchemy_query_STRUCT]
sixes = session.query(Car).filter(Car.engine.CYLINDER.COUNT == 6)
# [END bigquery_sqlalchemy_query_STRUCT]
sixes2 = list(sixes)

# [START bigquery_sqlalchemy_query_getitem]
sixes = session.query(Car).filter(Car.engine["cylinder"]["count"] == 6)
# [END bigquery_sqlalchemy_query_getitem]
# fmt: on
sixes3 = list(sixes)


for sixes in [sixes1, sixes2, sixes3]:
    assert sorted(car.model for car in sixes) == ["Sebring", "Town and Counttry"]
```


so for step one I added an assert to catch the lack of this "dispatch" sooner:

```diff
diff --git a/lib/sqlalchemy/orm/attributes.py b/lib/sqlalchemy/orm/attributes.py
index 1098359ec..c25c8f197 100644
--- a/lib/sqlalchemy/orm/attributes.py
+++ b/lib/sqlalchemy/orm/attributes.py
@@ -892,6 +892,7 @@ class AttributeImpl:
         self.class_ = class_
         self.key = key
         self.callable_ = callable_
+        assert dispatch is not None
         self.dispatch = dispatch
         self.trackparent = trackparent
         self.parent_token = parent_token or self

```

from there I dont know what the problem is yet.   should take 5 minutes

**Reply by zzzeek** [reply]
OK I didnt confirm yet but I'm going to put 95% that it's this:

```
        def __getattr__(self, name):
            if name.lower() in self.expr.type._STRUCT_byname:
                return self[name]

```

let's see

**Reply by zzzeek** [reply, marked as answer]
yup, this is not allowing attr errors to pass, please change as follows:


```
        def __getattr__(self, name):
            if name.lower() in self.expr.type._STRUCT_byname:
                return self[name]
            else:
                raise AttributeError(name)

```

the InstrumentedAttribute hierarchy has its own `Comparator` that integrates with the one used by the datatype, and there is a `__getattr__()` scheme in play that returns attributes from both the `InstrumentedAttribute` and the `Comparator` depending on which one - which means if some code says something like "if hasattr(attribute, 'foo')", that has to return `False`, which you need the `AttributeError` here to achieve.  the specific place this AttributeError is needed to prevent the problem:

```
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/event/base.py(460)__get__()
-> if hasattr(obj, "_slots_dispatch"):
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/attributes.py(472)__getattr__()
-> return getattr(self.comparator, key)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/util/langhelpers.py(1335)__getattr__()
-> return self._fallback_getattr(key)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/properties.py(474)_fallback_getattr()
-> return getattr(self.__clause_element__(), key)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/sql/elements.py(1602)__getattr__()
-> return getattr(self.comparator, key)
> /home/classic/dev/sqlalchemy/test3.py(77)__getattr__()
-> raise AttributeError(name)

```

you can see at the top some deep inside of SQLAlchemy thing is asking, `if hasattr(some_attribute, "_slots_dispatch")` and it should get `False` here.   returning None, some kind of attribute-get loop is occurring (these are impossible to follow) and "dispatch" was coming back as None.

**Reply by CaselIT** [reply]
makes sense. the getattr loop is very hard to follow

**Reply by zzzeek** [reply]
in this case it's literally just a getattr returning None as the value rather than saying "I dont have that attribute", it's as simple as that.  that's the None dispatch.

**Reply by kiraksi** [reply]
Thank you so much @zzzeek and @CaselIT for your guidance and learned a lot about how to navigate SQLAlchemy from this conversation!
